chain/xrp/tx/tx.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	if tx.EncodeForSigning == nil {
+		return nil, errors.New("missing serialised XRP transaction")
+	}
+
+	// For k256 signing, XRP uses sha512[:32]
+	// https://github.com/XRPLF/xrpl-py/blob/17aad31f77452d30917b9e4544c9c87c274c0e3d/xrpl/core/keypairs/secp256k1.py#L95
+	digestSha512 := sha512.Sum512(tx.EncodeForSigning)
+	firstHalf := digestSha512[:32]
+
+	return []xc.TxDataToSign{firstHalf[:]}, nil
+
+}
+
+// AddSignatures adds a signature to Tx
+func (tx *Tx) AddSignatures(signatures ...xc.TxSignature) error {
+	if tx.TransactionSignature != nil {
+		return errors.New("transaction already signed")
+	}
+
+	for _, rsvBytes := range signatures {
+		r, s, err := DecodeEcdsaSignature(rsvBytes)
+		if err != nil {
+			return err
+		}
+
+		signature := ecdsa.NewSignature(&r, &s)
+		signatureBytes := signature.Serialize()
+		signatureHex := hex.EncodeToString(signatureBytes)
+		tx.XRPTx.TxnSignature = signatureHex
+		tx.TransactionSignature = append(tx.TransactionSignature, rsvBytes)
+	}
+
+	return nil
+}
+
+func DecodeEcdsaSignature(signature xc.TxSignature) (btcec.ModNScalar, btcec.ModNScalar, error) {
+	var err error
+	var r btcec.ModNScalar
+	var s btcec.ModNScalar
+	rsv := [65]byte{}
+	if len(signature) != 65 && len(signature) != 64 {
+		return r, s, errors.New("signature must be 64 or 65 length serialized bytestring of r,s, and recovery byte")
+	}
+	copy(rsv[:], signature)
+
+	// Decode the signature and the pubkey script.
+	rInt := new(big.Int).SetBytes(rsv[:32])
+	sInt := new(big.Int).SetBytes(rsv[32:64])
+
+	rBz := r.Bytes()
+	sBz := s.Bytes()
+	rInt.FillBytes(rBz[:])
+	sInt.FillBytes(sBz[:])
+	r.SetBytes(&rBz)
+	s.SetBytes(&sBz)
+	return r, s, err
+}
+
+// GetSignatures returns back signatures, which may be used for signed-transaction broadcasting
+func (tx *Tx) GetSignatures() []xc.TxSignature {
+	return tx.TransactionSignature
+}
+
+// Serialize returns the serialized tx
+func (tx Tx) Serialize() ([]byte, error) {
+	if tx.XRPTx == nil {
+		return []byte{}, errors.New("missing XRP transaction")
+	}
+
+	xrpTx := tx.XRPTx
+
+	if xrpTx.LastLedgerSequence == 0 {
+		return []byte{}, errors.New("missing last ledger sequence")
+	}
+
+	if xrpTx.TxnSignature == "" {
+		return []byte{}, errors.New("missing transaction signature")
+	}
+
+	amountLargeNumber := new(big.Int)
+	amountLargeNumber.SetString(xrpTx.Amount.StringValue, 10)
+
+	divisor := big.NewInt(1000)
+
+	XRPAmount := new(big.Int)
+	XRPAmount.Div(amountLargeNumber, divisor)
+
+	result := make(map[string]interface{})
+	result["Account"] = string(xrpTx.Account)
+	result["Amount"] = XRPAmount.String()
+	result["Destination"] = string(xrpTx.Destination)
+	result["Fee"] = xrpTx.Fee
+	result["Flags"] = xrpTx.Flags
+	result["LastLedgerSequence"] = xrpTx.LastLedgerSequence
+	result["Sequence"] = xrpTx.Sequence
+	result["SigningPubKey"] = xrpTx.SigningPubKey
+	result["TransactionType"] = string(xrpTx.TransactionType)
+	result["TxnSignature"] = xrpTx.TxnSignature
```

Comment:
Duplicated.  I don't like how this rendering is done in both `tx` and `builder` modules.  I don't think it needs to be done in `builder` module, so i'd prefer to refactor to only do it in `tx` module.